src/cascadia/PublicTerminalCore/HwndTerminalAutomationPeer.cpp

```diff
@@ -0,0 +1,159 @@
+    {
+        if (const auto keyEventChar{ gsl::narrow_cast<wchar_t>(charCode) }; IsReadable({ &keyEventChar, 1 }))
+        {
+            _keyEvents.emplace_back(keyEventChar);
+        }
+    }
+}
+
+// Implementation of IRawElementProviderSimple::get_PropertyValue.
+// Gets custom properties.
+IFACEMETHODIMP HwndTerminalAutomationPeer::GetPropertyValue(_In_ PROPERTYID propertyId,
+                                                            _Out_ VARIANT* pVariant) noexcept
+{
+    pVariant->vt = VT_EMPTY;
+
+    // Returning the default will leave the property as the default
+    // so we only really need to touch it for the properties we want to implement
+    if (propertyId == UIA_ClassNamePropertyId)
+    {
+        // IMPORTANT: Do NOT change the name. Screen readers like may be dependent on this being "WpfTermControl".
+        pVariant->bstrVal = SysAllocString(L"WPFTermControl");
+        if (pVariant->bstrVal != nullptr)
+        {
+            pVariant->vt = VT_BSTR;
+        }
+    }
+    else
+    {
+        // fall back to shared implementation
+        TermControlUiaProvider::GetPropertyValue(propertyId, pVariant);
+    }
+    return S_OK;
+}
+
+// Method Description:
+// - Signals the ui automation client that the terminal's selection has changed and should be updated
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void HwndTerminalAutomationPeer::SignalSelectionChanged()
+{
+    UiaTracing::Signal::SelectionChanged();
+    LOG_IF_FAILED(UiaRaiseAutomationEvent(this, UIA_Text_TextSelectionChangedEventId));
+}
+
+// Method Description:
+// - Signals the ui automation client that the terminal's output has changed and should be updated
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void HwndTerminalAutomationPeer::SignalTextChanged()
+{
+    UiaTracing::Signal::TextChanged();
+    LOG_IF_FAILED(UiaRaiseAutomationEvent(this, UIA_Text_TextChangedEventId));
+}
+
+// Method Description:
+// - Signals the ui automation client that the cursor's state has changed and should be updated
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void HwndTerminalAutomationPeer::SignalCursorChanged()
+{
+    UiaTracing::Signal::CursorChanged();
+    LOG_IF_FAILED(UiaRaiseAutomationEvent(this, UIA_Text_TextSelectionChangedEventId));
+}
+
+void HwndTerminalAutomationPeer::NotifyNewOutput(std::wstring_view newOutput)
+{
+    // Try to suppress any events (or event data)
+    // that is just the keypress the user made
+    auto sanitized{ Sanitize(newOutput) };
+    while (!_keyEvents.empty() && IsReadable(sanitized))
+    {
+        if (til::toupper_ascii(sanitized.front()) == _keyEvents.front())
+        {
+            // the key event's character (i.e. the "A" key) matches
+            // the output character (i.e. "a" or "A" text).
+            // We can assume that the output character resulted from
+            // the pressed key, so we can ignore it.

```

Comment:
There's no way this works with Unicode, right? ._.

Reply:
Probably not, but that should be fine. We're really just using this to suppress key events. So, when you press the <kbd>A</kbd> key, normally you'd hear "A" from pressing the key and "A" from rendering the key. If we suppress that second event, it sounds/feels more natural.